src/Solver/Dynamic/Ida.cpp

```diff
@@ -426,6 +426,17 @@ namespace AnalysisManager
       checkOutput(retval, "IDASetMaxNumSteps");
 
       // Set up linear solver
+
+      if (JacobianMatB_ != nullptr || linearSolverB_ != nullptr)
+      {
+        // guard against repeated initialization
+        SUNLinSolFree(linearSolverB_);
+        SUNMatDestroy(JacobianMatB_);
+
+        linearSolverB_ = nullptr;
+        JacobianMatB_  = nullptr;
+      }
```

Comment:
Must make these separate checks.

CC @Steven-Roberts

Reply:
The two are allocated independently, so it is possible to have one and not the other.

Reply:
I also believe there are SUNDIALS methods to check if the matrix and/or linear solver are instantiated. @Steven-Roberts would know more about that than me.

Reply:
There is a `deleteBackwardSimulation` function which can free these objects. I think this is intended to be called after `initializeBackwardSimulation` and the adjoint solution has been computed, but correct me if I'm wrong @pelesh. If we don't want to rely on user code calling `deleteBackwardSimulation`, we can call that function at the top of `initializeBackwardSimulation`.

Reply:
After looking at the code, this is not an easy fix. I suggest creating a separate issue and removing this change altogether from this PR.
```suggestion
```

Reply:
Thanks, @superwhiskers, for pointing to this issue.

Reply:
>I also believe there are SUNDIALS methods to check if the matrix and/or linear solver are instantiated. @Steven-Roberts would know more about that than me.

I'm not aware of one. Checking if it is null is fine, although if it's going to be freed if non-null, you can directly call the SUNDIALS free/destroy method without wrapping it in an if statement.

> i'm pretty sure i tried this already and since it frees three other vectors, i got a null pointer dereference or something

I see. We could merge the `configureAdjoint` and `initializeBackwardSimulation` functions to fix that. Another option is not to free and reallocate the matrix and linear solver each backward integration. In `initializeBackwardSimulation` we can only create the matrix and linear solver if they are null.

Reply:
I think this is the best solution in this case.

Reply:
It would be handy to have an example that performs several backward integrations so we can settle on what functions get called in user code and where objects need to be allocated/freed.

Reply:
In [29d3221](https://github.com/ORNL/GridKit/pull/134/commits/29d322119c87f190cfcca44f32faef2dd66fcda4).